boost/text/text_view.hpp

```diff
@@ -0,0 +1,340 @@
+        sequence which is not. */
+    struct text_view
+    {
+        using iterator = char const *;
+        using const_iterator = char const *;
+        using reverse_iterator = detail::const_reverse_char_iterator;
+        using const_reverse_iterator = detail::const_reverse_char_iterator;
+
+        /** Default ctor.
+
+            This function is constexpr.
+
+            \post data() == nullptr && size() == 0 */
+        constexpr text_view () noexcept :
+            data_ (nullptr),
+            size_ (0)
+        {}
+
+        /** Constructs a text_view from a null-terminated C string.  The UTF-8
+            encoding is checked only at the beginning and end of the string,
+            to prevent slicing of code points.  To fully check the encoding,
+            use checked_encoding().
+
+            This function is constexpr in C++14 and later.
+
+            \pre strlen(c_str) <= max_size()
+            \throw std::invalid_argument if the ends of the string are not valid UTF-8.
+            \post data() == c_str && size() == strlen(c_str) */
+        BOOST_TEXT_CXX14_CONSTEXPR text_view (char const * c_str) :
+            data_ (c_str),
+            size_ (detail::strlen(c_str))
+        {
+            assert(detail::strlen(c_str) <= max_size());
+            if (!utf8::starts_encoded(begin(), end()))
+                throw std::invalid_argument("The start of the given string is not valid UTF-8.");
+            if (!utf8::ends_encoded(begin(), end()))
+                throw std::invalid_argument("The end of the given string is not valid UTF-8.");
+        }
+
+        /** Constructs a text_view from an array of char.  The UTF-8 encoding
+            is checked only at the beginning and end of the string, to prevent
+            slicing of code points.  To fully check the encoding, use
+            checked_encoding().
+
+            This function is constexpr in C++14 and later.
+
+            \throw std::invalid_argument if the ends of the string are not valid UTF-8.
+            \pre len >= 0
+            \post data() == c_str && size() == len */
+        BOOST_TEXT_CXX14_CONSTEXPR text_view (char const * c_str, int len) :
+            data_ (c_str),
+            size_ (len)
+        {
+            assert(0 <= len);
+            if (!utf8::starts_encoded(begin(), end()))
+                throw std::invalid_argument("The start of the given string is not valid UTF-8.");
+            if (!utf8::ends_encoded(begin(), end()))
+                throw std::invalid_argument("The end of the given string is not valid UTF-8.");
+        }
+
+        /** Constructs a text_view from a null-terminated C string, without
+            any check of UTF-8 encoding.
+
+            This function is constexpr in C++14 and later.
+
+            \pre strlen(c_str) <= max_size()
+            \post data() == c_str && size() == strlen(c_str) */
+        BOOST_TEXT_CXX14_CONSTEXPR
+        text_view (char const * c_str, utf8::unchecked_t) noexcept :
```

Comment:
Should there be another set of text_view ctor overloads that check the entire view? E.g. take "utf8::checked_t" or similar. This avoids the syntactic overhead of constructing with unchecked_t and then calling checked_encoding.

Reply:
I like that overhead, in that it stands out visually.  This is of course entirely a matter of taste.